packages/blob/src/api.ts

```diff
@@ -268,20 +265,104 @@ export async function requestApi<TResponse>(
   const apiResponse = await retry(
     async (bail) => {
       let res: Response;
+      let bodyLength: number | undefined;
+      let body: BodyInit | undefined;
+
+      if (
+        init.body &&
+        // 1. For upload progress we always need to know the total size of the body
+        // 2. In development we need the header for put() to work correctly when passing a stream
+        (commandOptions?.onUploadProgress || shouldUseXContentLength())
+      ) {
+        bodyLength = computeBodyLength(init.body);
+      }
+
+      if (init.body) {
+        if (commandOptions?.onUploadProgress) {
+          if (supportsRequestStreams) {
+            // We transform the body to a stream here instead of at the call site
+            // So that on retries we can reuse the original body, otherwise we would not be able to reuse it
+            const stream = await toReadableStream(init.body as PutBody);
+
+            let loaded = 0;
+
+            const chunkTransformStream = createChunkTransformStream(
+              CHUNK_SIZE,
+              (newLoaded: number) => {
+                loaded += newLoaded;
+                const total = bodyLength ?? loaded;
+                const percentage = Number(((loaded / total) * 100).toFixed(2));
+
+                // Leave percentage 100 to end of request
+                if (percentage === 100) {
+                  return;
+                }
+
+                commandOptions.onUploadProgress?.({
```

Comment:
It seems required, probably because the call is too far away for TS to be sure it will still be here even after the if? 
![CleanShot 2024-10-24 at 14 16 13@2x](https://github.com/user-attachments/assets/4472bd67-94b0-460f-bfbb-196f84b1d278)

I've added a local variable to keep a reference to it in 59e01f6192cff9f9e74ed61ff2ba8db63e427260

Reply:
Ah I see. It's probably because it's within a callback